### DISCUSSION TITLE
How do I obtain the the signature file for the 2.0.28 source tabrall?

### DISCUSSION THREAD

**Discussion by loqs** [sqlalchemy/sqlalchemy#11138] [Usage Questions]
Since [pypi removed pgp support](https://blog.pypi.org/posts/2023-05-23-removing-pgp/) the signature files for new releases have not been available from pypi. Or any alternate way to link the current release to the key `83AF7ACE251C13E6BB7DEFBD330239C1C4DAFEE1` `Michael Bayer <mike_mp@zzzcomputing.com>` Related [fix-reproducibility-of-packages-broken-by-pypi-removing-signature-files](https://archlinux.org/todo/fix-reproducibility-of-packages-broken-by-pypi-removing-signature-files/). A local copy of the signature for the 1.4.44 tarball can be used to continue verifying it but does not help with respect to newer releases.

**Comment by zzzeek** [comment]
I dont sign the packages anymore.  pypa sends me an obnoxious email if I generate a signature and upload it to them.  tell them to turn off the obnoxious email and I can start pushing signature files up there again if they give a way to get them.

**Reply by loqs** [reply]
Thanks for the fast response.  The pypa situation is unfortunate and I have no influence over it. Would you consider uploading the signature file to https://github.com / https://www.sqlalchemy.org or another host. Another alternative would be signing the git releases either the commit or the tag. @jelly @dvzrv @anthraxx other possibilities?

**Reply by CaselIT** [reply]
Is not using the signature not an option?
I'm guessing pypa removed it for a good reason, so I don't think continuing to require it is a great solution.

**Reply by dvzrv** [reply]
Hi! :wave: 

From a distribution standpoint we are not very happy about the PyPI decision to remove integration for OpenPGP signatures as it left us with a few of these unclear situations.

We have since [moved away from relying on sdist tarballs hosted on PyPI](https://rfc.archlinux.page/0020-sources-for-python-packaging/) (also because the sdist format lacks a definition and we often had issues with it all across, e.g. due to missing tests, etc.).

FWIW, it is possible to add signed tarballs to [the GitHub releases of this project](https://github.com/sqlalchemy/sqlalchemy/releases) - it seems you are already (automatically?) adding many different types of wheels as artifacts.
For us this would be the ideal solution, as we would like to maintain a trust path going forward.
Please note, that we would like a trust path to exist between releases. This means that e.g. a cryptographic trust path exists between one OpenPGP certificate and the next (by third-party certifications or other means, such as adding relevant OpenPGP fingerprints to a document in this project in a signed commit).
In case you have questions in regards to that, feel free to reach out to me!

That said, I am not here to tell you how to run your project and I acknowledge, that you may find this all to be too much overhead.
In case you do not want to continue signing release tarballs at all anymore, it would still be really awesome to get a cleartext signed message from @zzzeek so that we can with good conscience remove this verification scheme from our packaging setup.
This could be done with the following (and just appending the resulting `message.asc` file).

```shell
gpg --armor --default-key 83AF7ACE251C13E6BB7DEFBD330239C1C4DAFEE1 --output message.asc --clearsign <<< "I hereby acknowledge that the sqlalchemy project no longer signs its source tarballs using OpenPGP."
```

**Reply by CaselIT** [reply]
> FWIW, it is possible to add signed tarballs to [the GitHub releases of this project](https://github.com/sqlalchemy/sqlalchemy/releases) - it seems you are already (automatically?) adding many different types of wheels as artifacts.

You mean adding one on top of the automatically generated one whrn creating a release?
Since the sdist tar also includes rendered documentation that's not in the automaic tar file created by GitHub

**Reply by zzzeek** [reply]
everything on GH is automated so we'd probably have to create an machine-owned GPG key to also automate the signing

**Reply by CaselIT** [reply]
> everything on GH is automated so we'd probably have to create an machine-owned GPG key to also automate the signing

you can upload a file to an existing release. This seems to be easily doable in a bash script too using the `gh` command: `gh release upload` https://cli.github.com/manual/gh_release_upload

So the workflow could be:
- a release is created manually on github, this triggers the wheel pipeline that does its thing (same as now)
- mike creates the sdist and uploads it to pypi (same as now)
- mike signs the sdist and uploads it (and the signature? not sure if it's another file) to the release using the gh command. (this is new)

Would something like this work?

**Reply by dvzrv** [reply]
Sorry for the rather late reply. We've been quite busy rebuilding everything for Python 3.12 the past weeks... (plus dayjob).

>  Would something like this work?

Yes (from our side). However, we would rather like to rely on just auto-generated source tarballs (even for those signatures can be created and be attached to a release!)

Alternatively, and this may be much easier for you: Can you start relying on signed tags instead? Or is there a particular reason to auto-create (non-annotated/ unsigned) tags?
Do note that if you are considering automated signed tags for this (with a softkey somewhere in a CI system), please don't bother: We would not want to rely on them.

I'm bringing up signed tags in this context as they would represent a more transparent source than sdist tarballs.

**Reply by CaselIT** [reply]
The source tarball published to pypi is not just a tag of the repo, but includes the rendered docs etc.
That's currently manually made and uploaded to pypi by mike to ensure that the docs came out ok and everything else is all right.

If for your purpose a git tag is enough, if I'm not mistaken that's also generated locally then uploaded by mike so I think setting up a signing key with github to generate signed tags could be an option, but let's wait for his input on this. cc @zzzeek

**Reply by zzzeek** [reply]
still not really into this as this was a pypa decision.  is SQLAlchemy the only Python package you use?   (clearly not since you have to have a DBAPI driver also).    Are you posting issues on every Python dependency for this ?    What decisions are being made on those packages?   This really seems like an issue above us that should be solved by pypa or some other third party.
